### ISSUE TITLE
commit 67ed0bc (Issue161)

### ISSUE THREAD

**Issue by Ultima1252** [ellzey/libevhtp#201]
After this commit seafile-client no longer works. I know this isn't seafile, however I assume there is some desire to have programs to run on newer versions of libevhtp.

 This is the error message seafile gives. Not sure if  this will help.

libcurl failed to GET https://domain.com/seafhttp/repo/7bf6c298-4b05-4ee2-92eb-f60d67500791/permission-check/?op=download&client_id=297e89d4b46c5308ea41b71ba5bcd128de8cb4ef&client_name=someuser%40somehost: Transferred a partial file.

 Let me know if theres any more information needed to help fix this issue.

**Comment by ellzey** [comment]
Can you post the verbose flag in curl for me? Thanks in advance!

**Comment by ellzey** [comment]
> however I assume there is some desire to have programs to run on newer versions of libevhtp.

I have always tried not to break things. If something is going to be deprecated; I give fair warning with the __attribute__(deprecated), and that's only happened once... And that was today

**Comment by Ultima1252** [comment]
Ah, that was actually the error in seafile.log. When I try curl -v "%s" it just fails with bad request, and that is on a working version. This is the entire entry for it, but I don't think this will help at all.

[11/26/15 23:34:36] clone-mgr.c(812): Transition clone state for 7bf6c298 from [fetch] to [fetch].
[11/26/15 23:34:36] http-tx-mgr.c(3946): Download with HTTP sync protocol version 1.
[11/26/15 23:34:36] http-tx-mgr.c(975): Transfer repo '7bf6c298': ('normal', 'init') --> ('normal', 'check')
[11/26/15 23:34:36] http-tx-mgr.c(678): libcurl failed to GET https://domain.com/seafhttp/repo/7bf6c298-4b05-4ee2-92eb-f60d67500791/permission-check/?op=download&client_id=297e85d4b84c5308ea41b71da7bcd988de8cb6ef&client_name=user%40host: Couldn't connect to server.
[11/26/15 23:34:36] http-tx-mgr.c(3952): Download permission denied for repo 7bf6c298 on server domain.com.
[11/26/15 23:34:36] http-tx-mgr.c(975): Transfer repo '7bf6c298': ('normal', 'check') --> ('error', 'finished')

Unfortunately, I cant figure out how to start debugging for seafile and it is not documented anywhere. The variable in the source seems to do nothing at all when adding it to the config. =/

 Also, I wasn't sure if features were being removed that were causing the break, this must just be code that just needs more testing. Thanks for making that clear =]

Ill try and contact seafile and see if I can get a clear answer about this debugging variable, and where it may go (the *.conf is what I assume) >.<

**Comment by ellzey** [comment]
I can't help you debug code I have no practical knowledge of, but giving it a look-over; `http-tx-mgr` is part of the `daemon` - which does not seem to use evhtp. `seafile-server` is where evhtp is used.

Whether they are associated with each other, I haven't a clue.

**Comment by ellzey** [comment]
My best guess is, look at the error 'Couldn't connect to server.'. Are you actually making a request to domain.com? If so, that's your problem.

**Comment by Ultima1252** [comment]
All I'm sure about is building  building seafile with libevhtp on commit b1cef41, everything works fine. Then test it again on commit 67ed0bc and it nolonger is able to download due with the error transferred partial file. That is the only change between tests. Now I created a patch for 1.2.11 with reverting the patch on commit 67ed0bc, and everything works perfectly. Sorry I can't be of more help =/

**Comment by ellzey** [comment]
Read the response I made here, it was my fault. https://github.com/haiwen/seafile/issues/1119

**Comment by Ultima1252** [comment]
ah yes, I forgot to mention, I already have a patch that removes EV_WRITE on each of my builds. So with it removed I am having this issue. That is an issue that is currently affecting seafile, after reviewing it, I do believe this maybe on implementation on the seafile side, but is more than likely a separate issue.

**Comment by ellzey** [comment]
Yes, but that breaks pipelined request processing. Read the reply I wrote on the seafile issue above.

**Comment by Ultima1252** [comment]
The issue has become a bit clearer to me now. Thanks for your help. =] Ill start looking more at the seafile implementation.

**Comment by ellzey** [comment]
This needs to be kept open while I fix seafile for my fixes that resulted in bad behavior if not done in the exact order i put up there. On top of that, it wasn't made clear in the changelog that this might be an issue.

It's one of those "if a butterfly flaps its wings.." Situations

**Comment by Ultima1252** [comment]
Created a patch to fix seafile request_resume before reply, this has fixed the my other issue I posted, so will close as it is a seafile issue and not libevhtp. However I still think this is a separate issue as with the fix, the client is still providing the error message in my previous comment. Tested with the patch reverted works as it should. Is this a similar change for keep_alive connections that may have broke things?

**Comment by Ultima1252** [comment]
This is unfortunate but, I think I may have found a third issue, or my current seafile patch is not quite complete. Doing a clean update to 1.2.10 with the seafile patch fixes the seafile-client. Upgrading to 1.2.11 clean still breaks the client, but reverting 67ed0bc keeps seafile-client running strong. In fact it works better than ever. Will work on a patch for 67ed0bc, however first want to address this new issue.

 The new issue seems to be downloading from the web interface. Which is what my first issue was for. (The one I closed) My "gut" feeling is that its on the seahub side so I will investigate there, but basically my old temp fix, which would break pipelined requests and cause some havoc for the client fixed the web interface. Ill report back my findings.

**Comment by ellzey** [comment]
I really don't care about client stuff: please read the wiki https://github.com/ellzey/libevhtp/wiki/Mark-on-evhtp-client-API on this topic.

**Comment by Ultima1252** [comment]
Well, I just found the problem I know you just posted this but... the issue looks like something you were working on anyway.

--- evhtp.c.orig        2015-11-29 01:12:38 UTC
+++ evhtp.c
@@ -1673,12 +1673,10 @@ check_proto:
                                          evhtp_header_new("Connection", "close", 0,
 0));
             }

-#if 0
-            if (!out_len && !evhtp_header_find(request->headers_out, "Content-Lengt
h")) {
+            if (!evhtp_header_find(request->headers_out, "Content-Length")) {
                 evhtp_headers_add_header(request->headers_out,
                                          evhtp_header_new("Content-Length", "0", 0,
 0));
             }
-#endif

             break;
         case EVHTP_PROTO_10:


this patch makes it work perfectly

**Comment by ellzey** [comment]
can you do a pull request?
